app/uk/gov/hmrc/gform/sharedmodel/VariadicFormData.scala

```diff
@@ -213,8 +216,103 @@ object VariadicFormData {
     // properly classify data to VariadicValues
     val multiValueIds: Set[BaseComponentId] = formModel.allMultiSelectionIds.map(_.baseComponentId)
 
+    val dataGroupBy = data.groupBy(_._1.baseComponentId)
+
+    val allGroupFormComponentsMap = formModel.allFormComponents.collect {
+      case fc @ IsGroup(group) => fc.baseComponentId -> (fc.baseComponentId +: group.fields.map(_.baseComponentId))
+    }.toMap
+
+    def isPureExists(ids: List[ModelComponentId]): Boolean =
+      ids.exists { r =>
+        r match {
+          case Pure(iPure(BaseComponentId(_))) => true
+          case _                               => false
+        }
+      }
+
+    def isAtomicExists(ids: List[ModelComponentId]): Boolean =
+      ids.exists { r =>
+        r match {
+          case Atomic(iPure(BaseComponentId(_)), _) => true
+          case _                                    => false
+        }
+      }
+
+    def getFormComponentId(baseComponentId: BaseComponentId): Option[BaseComponentId] =
+      allGroupFormComponentsMap.find(_._2.contains(baseComponentId)).map(_._1)
```

Comment:
Same here. use of _1  and _2 in tuples makes it hard to read. Can we use case expressions instead please?